# 01 Importing libraries

In [1]:
# Importing libraries
import pandas as pd
import numpy as np
import os

# 02 Importing data

In [2]:
# Importing dataset

df = pd.read_csv(r'C:\Users\abiha\OneDrive\Documents\IC Project Folder\2. Data\Original Data\orders.csv', index_col = False)

In [3]:
# Set path

orderpath = r'C:\Users\abiha\OneDrive\Documents\IC Project Folder\2. Data\Original Data\orders.csv'

In [7]:
# Import orders df

df_ords = pd.read_csv(os.path.join(orderpath), index_col = False)

In [5]:
# Set product df path

productpath = r'C:\Users\abiha\OneDrive\Documents\IC Project Folder\2. Data\Original Data\products.csv'

In [6]:
# Import products df

df_prods = pd.read_csv(os.path.join(productpath), index_col = False)

In [8]:
# Check frequencies of 'days_since_prior_order' column

df_ords['days_since_prior_order'].value_counts(dropna = False)

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

In [9]:
# Set path for departments df

departmentpath = r'C:\Users\abiha\OneDrive\Documents\IC Project Folder\2. Data\Original Data\departments.csv'

In [10]:
# Import departments df

df_dep = pd.read_csv(os.path.join(departmentpath), index_col = False)

In [13]:
# Check departments df

df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

# 03 Exercise 4.4 wrangling

In [14]:
# Transpose df_dep dataframe
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [15]:
# Save changes

df_dep_t = df_dep.T

In [16]:
# Check output

df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [17]:
# add index to df_dep dataframe
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [18]:
# Change header
new_header = df_dep_t.iloc[0]

In [19]:
# Save changes

df_dep_t_new = df_dep_t[1:]

In [20]:
# Check output

df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [21]:
# assign new header

df_dep_t_new.columns = new_header

In [22]:
# Check output

df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [23]:
# Create data dictionary using df_dep
data_dict = df_dep_t_new.to_dict('index')

In [24]:
# Check data dictionary

data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [25]:
# Check prdocucts df

df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [26]:
# Create df for snacks department

df_snacks =  df_prods[df_prods['department_id']==19]

In [27]:
# Check output for snacks df

df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

# 04 Data Wrangling & Subsetting Task 4.4

### 2. Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.

In [30]:
# Check orders df 

df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [32]:
# changing data type of order_id column
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [33]:
# changing data type of user_id column
df_ords['user_id'] = df_ords['user_id'].astype('str')

### 3. Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the data frame.

In [34]:
# Changing order_dow to order_day_of_week WITHOUT overwriting the dataframe
df_ords.rename(columns = {'order_dow' : 'order_day_of_week'})

order_id user_id eval_set  order_number  order_day_of_week  \
0        2539329       1    prior             1                  2   
1        2398795       1    prior             2                  3   
2         473747       1    prior             3                  3   
3        2254736       1    prior             4                  4   
4         431534       1    prior             5                  4   
...          ...     ...      ...           ...                ...   
3421078  2266710  206209    prior            10                  5   
3421079  1854736  206209    prior            11                  4   
3421080   626363  206209    prior            12                  1   
3421081  2977660  206209    prior            13                  1   
3421082   272231  206209    train            14                  6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

### 4. Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.

In [37]:
# showing order counts for each hour of the day
df_ords['order_hour_of_day'].value_counts(dropna = False)

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_hour_of_day, dtype: int64

#### Between 10-11am is the busiest hour for placing orders with 288,418 

### 5. Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.

In [38]:
# bringing up data dictionary previously created
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [39]:
# Showing meaning of 4 in "department_id" column
print(data_dict.get('4'))

{'department': 'produce'}


#### Value of 4 in "department_id" column is for 'produce'

### 6. The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.

In [40]:
# Creating subset of orders with a department_id relating to breakfast item 
breakfast_item_sales = df_prods[df_prods['department_id'] == 14]

In [41]:
breakfast_item_sales

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

### 7. They’d also like to see details about products that customers might use to throw dinner parties. Your task is to find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.

In [42]:
# Creating a subset including all orders with department_id's relating to alcohol, deli, beverages and meat/seafood
dinner_party_sales = df_prods.loc[df_prods['department_id'].isin([5,20,7,12])]

In [43]:
# Checking subset

dinner_party_sales

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

### 8. It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?

#### The last dataframe 'dinner_party_sales' has 7650 rows

### 9. Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.

In [44]:
# creating subset with all information from user_id 1
user_id_1 = df_ords[df_ords['user_id'] == 1]

In [45]:
# Checking output

user_id_1

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

In [46]:
# changing user_id data type back to integer so it can be analysed
df_ords['user_id'] = df_ords['user_id'].astype('int')

In [47]:
# creating subset with all information from user_id 1

user_id_1 = df_ords[df_ords['user_id'] == 1]

In [48]:
# Checking output

user_id_1

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   
5   3367565        1    prior             6          2                  7   
6    550135        1    prior             7          1                  9   
7   3108588        1    prior             8          1                 14   
8   2295261        1    prior             9          1                 16   
9   2550362        1    prior            10          4                  8   
10  1187899        1    train            11          4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

### 10. You also need to provide some details about this user’s behavior. What basic stats can you provide based on the information you have?

#### I can provide information on how many orders have been placed, which days of the week they occurred, what time of the day and how many days since their prior order. I can also give them order_id numbers.

### 12. Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder.

In [50]:
df_ords.to_csv(os.path.join(r"C:\Users\abiha\OneDrive\Documents\IC Project Folder\2. Data\Prepared Data", 'orders_wrangled.csv'))

###  13. Export the df_dep_t_new dataframe as “departments_wrangled.csv” in your “Prepared Data” folder so that you have a “.csv” file of your departments data in the correct format.

In [51]:
df_dep_t_new.to_csv(os.path.join(r"C:\Users\abiha\OneDrive\Documents\IC Project Folder\2. Data\Prepared Data", 'departments_wrangled.csv'))